k2: refactor zserial to allow use of multiple serial ports

Risca opened the pull request:
Systests for zserial has not been converted. The plan is to make this backwards compatible.

This changes so that `SerialExtension` is instantiated on `SERIAL_PORT_IDS` instead of `SUT`.

Example config file:
```yaml
serial.port.ids:
  - red
  - black

serial.port.red.device: /dev/ttyUSB0
serial.port.red.enabled: true
serial.port.red.baudrate: 9600
serial.port.black.device: /dev/ttyUSB1
serial.port.black.enabled: false
serial.port.black.cc.enabled: false

suts.ids:
  - box

suts.box.serial.ports:
  - black
  - red
```

Example use:
```python
from xmodem import XMODEM1k

from zaf.component.decorator import requires

@requires(serial='RawSerialPort')
@requires(exec='Exec', can=['serial'])
def test_xmodem_recv(serial,exec):
    with open('/tmp/foo.bar', 'wb') as f:
        exec.send_line('sz -X --1k -q /proc/cpuinfo', prefix_output=False, endmark='Give your local XMODEM receive command now.')
        with serial.suspended() as ser:
            def getc(size, timeout=1):
                return ser.read(size) or None

            def putc(data, timeout=1):
                return ser.write(data)

            modem = XMODEM1k(getc, putc)
            modem.recv(f)

@requires(connection='SerialPort', can=['red'])
@requires(serial='RawSerialPort', uses=['connection'])
@requires(exec='Exec', can=['serial'], uses=['connection'])
def test_xmodem_send(serial,exec):
    with open('/proc/version', 'rb') as f:
        exec.send_line('rz -X -q -y /tmp/foo', prefix_output=False, endmark='rz: ready to receive /tmp/foo')
        with serial.suspended() as ser:
            def getc(size, timeout=1):
                return ser.read(size) or None

            def putc(data, timeout=1):
                return ser.write(data)

            modem = XMODEM1k(getc, putc)
            modem.send(f)
```

Risca commented:
I plan to start working on making it backwards compatible, but I wanted to push what I had so that I could get some early feedback. Maybe my approach is entirely wrong? :shrug:

per-bohlin commented on k2/addons/k2-sutevents/sutevents/suteventslog.py:
```diff
@@ -107,14 +109,18 @@ def __init__(self, config, instances):
 
-    @sequential_dispatcher(
-        [LOG_LINE_RECEIVED, SUT_RESET_EXPECTED, SUT_RESET_NOT_EXPECTED], entity_option_id=SUT)
+    @sequential_dispatcher([SUT_RESET_EXPECTED, SUT_RESET_NOT_EXPECTED], entity_option_id=SUT)
```
I don't understand how this can work now when you have removed the LOG_LINE_RECEIVED. It will no longer process log-lines. Is this change really intentionally? If yes, what am I missing? How will it work?

per-bohlin replied:
I now I see, you register the LOG_LINE_RECEIVED above on lines 114-115. Never mind above comment.

per-bohlin commented:
In general I think it is a good solution. However it makes a couple of existing problems more obvious.

We are lacking a proper abstraction for processing logs created by the SUT. 
The introduction of multiple serial connections high-lights this already existing problem.

The intended abstraction between a log-producer and log-consumer is the LOG_LINE_RECEIVED event. 
Unfortunately, the abstraction was not fully realized and a shortcut was made and the serial component/plugin 
became the de-facto log-producer without any abstraction.

You can see it in that the old code makes references to the serial plugin through
SERIAL_ENABLED in suteventslog, to determine if a log-producer is active.
Ideally, the serial plugin should have registered itself as a log-producer on the sut.
The suteventslog could then look at that property.

The original intention was that other log-producers would be possible in the future (telnet, socket, etc).

Now with the introduction of multiple log-producers in the form of multiple serial ports, 
we need a mechanism to tell the suteventslog what log-source(s)/log-producer(s) to care about, monitor.
Essentially, what log-producing entities to listen to for the LOG_LINE_RECEIVED event.

If I understand the implementation correct, by default, SutEvents component and suteventslog will
listen on all ports on the sut.

SutEvents.wait_for_log_line offer the option to only listen to a specific port but by default will
listen to all ports on the sut.

It that is the case, that is the correct behavior - just the wrong abstraction.

I would recommend that we introduce the Log-producer/source concept. A serial port ID should
have the configuration option to make it a log-source or not. Default could be true and
have the same name/ID as the port-ID with a unique prefix.

Lets say we have both telnet and serial log-sources. If both a serial connection and telnet connection
is called "primary", it would result in conflicting log-IDs. Therefore it would be good
if the serial added itself as a log-source as "serial-primary" or similar.

So by introducing the list-property log.entities, log.sources or similarly named on the sut,
the serial plugin can add each serial port that has been configured as a log-source to that list.

The suteventslog would use sut.log.entities/sources as event entities to listen to.

That would remove the dependency between suteventslog and the serial addon.

The user could then specify the log-entity/source for SutEvents.wait_for_log_line if
they did not want to monitor all sources.

With the introduction of multiple sources, I think we should have an additional method
that allows us to wait for a specific message on all sources.
SutEvents.wait_for_log_line_on_all(regex).

Risca commented:
> So by introducing the list-property log.entities, log.sources or similarly named on the sut,
> the serial plugin can add each serial port that has been configured as a log-source to that list.
> 
> The suteventslog would use sut.log.entities/sources as event entities to listen to.

Do you think this can be solved satisfactory with the existing zaf/k2 framework, or do they need to be updated as well? I think the config framework should be good enough, but I worry about order of instantiation. Ex.:
1. A Serial extension gets instantiated and registers its entity `serial-A` as a log producer
2. A SutEvents extension gets instantiated and starts to listen to `serial-A`
3. A second Serial extension gets instantiated and registers its entity `serial-B` as a log producer.

How do we solve this? I'll dig into how `k2-power` does it with the `AVAILABLE_POWER_METERS` list and see if that mechanism might be a good fit :smiley:

Risca commented:
> > So by introducing the list-property log.entities, log.sources or similarly named on the sut,
> > the serial plugin can add each serial port that has been configured as a log-source to that list.
> > The suteventslog would use sut.log.entities/sources as event entities to listen to.
> 
> Do you think this can be solved satisfactory with the existing zaf/k2 framework, or do they need to be updated as well? I think the config framework should be good enough, but I worry about order of instantiation. Ex.:
> 
> 1. A Serial extension gets instantiated and registers its entity `serial-A` as a log producer
> 2. A SutEvents extension gets instantiated and starts to listen to `serial-A`
> 3. A second Serial extension gets instantiated and registers its entity `serial-B` as a log producer.
> 4. The SutEvents extension we instantiated in step 2, doesn't receive any log lines from `serial-B` disappointed
> 
> How do we solve this? I'll dig into how `k2-power` does it with the `AVAILABLE_POWER_METERS` list and see if that mechanism might be a good fit

I did another round of thinking and realized that this won't be an issue since dispatchers will be registered after extensions are instantiated (if my understanding of the zaf framework is correct)

Risca commented:
Another thought that crossed my mind today was:
* should multiple serial ports per SUT be considered an "advanced" configuration, and focus should instead be on single-port SUT configuration? 

In this pull request I introduce additional complexity; not only for the extension code, but also for the _usage_ of the extension. When I was working on fixing so that the zserial systests would pass again, I called the old options (e.g. `--suts-box@serial-device`) "legacy options" since I thought this new way and was better - more modular, and more fine grained - but is it making it too hard to setup the most simple single port SUT setup? Single port SUT setup has to be the most common use case, right? If so, shouldn't the main focus be to make sure this use case is working properly? A bit of a philosophical question, maybe, but I want to have the right mindset when I continue to work on this :smiley:

per-bohlin commented:
A thing we should keep in mind is that the original zserial implementation made a couple of very naive and blatant assumptions:
* that each SUT can only have one serial device each at most (the problem you are trying to fix)
* that only SUTs could have serial devices

The second assumption is in my opinion even more restrictive. What if you have a serial device to e.g. test equipment that is not related to a SUT at all.

The changes you have started on will make it possible to decouple the notion of a serial device from the SUT all together. I think that separation is a very good for the plugin as a whole and make it easier to use serial devices in general.

But I agree that the old notation was quite nifty and simple.

What if the old notation is a short-hand that will add a serial device in the general serial.port index in the background?

The following configuration:

```
suts.box.serial.enabled: true
suts.box.serial.device: /dev/ttyUSB0
suts.box.serial.baudrate: 9600
```

would result result in the following configuration being added by the extension itself:

```
serial.port.ids:
  - box

serial.port.box.device: /dev/ttyUSB0
serial.port.box.enabled: true
serial.port.box.baudrate: 9600

suts.box.serial.ports:
  - box
```

That way the short-hand notation could still be used, but internally we only use the new configuration style. Which will work since the "legacy"/short-hand notation is automatically translated to the new notation as well.
Internally, the legacy/short-hand configuration notation would never be used since it is guaranteed to be doubled in the new format.

That is how I would do it.

Risca commented:
> What if the old notation is a short-hand that will add a serial device in the general serial.port index in the background?
> 
> The following configuration:
> 
> ```
> suts.box.serial.enabled: true
> suts.box.serial.device: /dev/ttyUSB0
> suts.box.serial.baudrate: 9600
> ```
> 
> would result in the following configuration being added by the extension itself:
> 
> ```
> serial.port.ids:
>   - box
> 
> serial.port.box.device: /dev/ttyUSB0
> serial.port.box.enabled: true
> serial.port.box.baudrate: 9600
> 
> suts.box.serial.ports:
>   - box
> ```
> 
> That way the short-hand notation could still be used, but internally we only use the new configuration style. Which will work since the "legacy"/short-hand notation is automatically translated to the new notation as well.
> Internally, the legacy/short-hand configuration notation would never be used since it is guaranteed to be doubled in the new format.

I like it :smile_cat: 
It's very similar to how I fixed backwards compatibility in my [other  branch](/Risca/opensourcelib/tree/topic-multi-connection-serial-log-lines), based on the changes in #34. Just needs a few adjustments and sanity checks so that `suts.box.serial.foo` and `serial.port.box.foo` are not both given at the same time.

Only, it might not be called "backwards compatibility" but simply "another way to configure serial ports". Still not sure how to name things here. I kinda like the old simple configuration, and it would be counterproductive to call those options "legacy". Legacy, in my mind, means something that will be phased out, removed.
Maybe just sort them as "sut (short) options"? E.g.
```python
from zseria.sut import SUT_SERIAL_PORTS, SUT_SERIAL_ENABLED, etc.
```

Risca commented:
I will keep this pull request as a draft until #34 (or equivalent) is merged

Risca commented:
I think this is ready for review now. I'll add some more tests later this week

Risca commented on k2/addons/k2-serial/zserial/serial.py:
```diff
@@ -123,27 +127,36 @@ def suspended(self):
+        ConfigOption(SUT_SERIAL_FILTERS, required=False),
+        ConfigOption(SUT_SERIAL_LOG_ENABLED, required=False),
+        ConfigOption(SUT_SERIAL_PROMPT, required=False),
+        ConfigOption(SUT_SERIAL_TIMEOUT, required=False),
```
Even if many of these options are not used by this extension, I had problems passing the systests without them listed here. I'm guessing it's because this extension gets disabled before the `SerialLogSourceExtension` gets a chance to add the necessary options, but I didn't investigate further

Risca replied:
Nope, it's not the load order. `SerialLogSourceExtension` has load order 91 and `SerialExtension` has load order 1000 (default for `@CommandExtension`)

Risca commented:
Added some unittests and some minor cleanups

Risca commented:
Any chance to get this merged soon?

per-bohlin commented:
Sorry, it slipped my mind. Completely forgot about it. I will not have time to look at it for at least two weeks. So go ahead and merge if it is blocking you.

Risca commented:
> Sorry, it slipped my mind. Completely forgot about it. I will not have time to look at it for at least two weeks. So go ahead and merge if it is blocking you.

That's ok. I'll merge this and we can rework it later on if needed
